Daily Readings Views Updates

vyvyvyThao opened the pull request:
# Description

Update the daily_readings_unit view so that it reuses the hourly_readings_unit view, fastening runtime for query; change the code for refreshing views so that hourly readings are always complete before daily readings are refreshed.

## Type of change

- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations
Test LG6 failed with message "Timeout of 15000ms exceeded". However, test succeeded when running only tests in readingsLineGroupQuantity file with the `npm run testsome` command.

Update: I am not seeing a timeout on testing so assume it not a real issue.

Commit:
add reading views

Commit:
update meter daily reading views and refresh reading views function

huss commented on src/server/services/refreshAllReadingViews.js:
```diff
@@ -9,12 +9,25 @@ const { log } = require('../log');
 const refreshDailyReadingViews = require('./refreshReadingViews').refreshReadingViews;
 const { refreshHourlyReadingViews } = require('./refreshHourlyReadingViews');
 
+/** 
```
I think my discussion of these changes was confusing. This is what OED wanted:

- Put the deprecated comments on refreshDailyReadingViews & refreshHourlyReadingViews.
- Remove the old code that is now called refreshAllReadingViewsOld and its export.

huss commented on src/server/services/refreshAllReadingViews.js:
```diff
@@ -9,12 +9,25 @@ const { log } = require('../log');
 
 	log.info('Refreshing All Reading Views...');
-	await Promise.all([refreshDailyReadingViews(), refreshHourlyReadingViews()]);
+	await refreshHourlyReadingViews();
```
Could you please add a comment on why this is now done sequentially with hourly first.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -344,8 +240,6 @@ hourly_readings_unit
 
 
 -- TODO Check if needed and when to use as not done for hourly.
```
Can you please put back the rest of the TODO comment that was there before:

```
-- With the index added in 3D readings, this should be consider as part of the decision
-- on if this is needed.
```

huss replied:
Now that the daily view depends on the hourly view, the hourly view must be declared first. This is why all the tests could not run on the PR. I moved this below hourly and the basic problem went away but there are still failed tests (see other comment).

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -96,126 +96,22 @@ so the line reading functions can use them both in the same way.
-	-- This sequence of joins takes the meter id to its unit and a unit.
-	INNER JOIN meters m ON r.meter_id = m.id)
+		h.meter_id AS meter_id,
+        avg(h.reading_rate) AS reading_rate,
```
The original indenting was done with tabs so can the spaces be replaced with them. VSC can do this for you.

Please make ave, max & min capitalized. I realize OED is not always doing this but it is best to put SQL commands as uppercase.

vyvyvyThao replied:
Discussed and decided to keep these lowercase

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -684,7 +579,7 @@ BEGIN
 		--  dr.reading_rate is the weighted average reading rate per hour over the day.
```
I'm not sure why this block is showing as a change as most seems the same. However, sum should be returned to SUM.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -96,126 +96,22 @@ so the line reading functions can use them both in the same way.
 CREATE MATERIALIZED VIEW IF NOT EXISTS
```
There needs to be a migration for this DB change. I can help with that if you want. I note it will involve dropping the view so it can be recreated.

huss commented on src/server/services/refreshAllReadingViews.js:
```diff
@@ -9,12 +9,25 @@ const { log } = require('../log');
+	log.info('...Views Refreshed!');
+}
 
 async function refreshAllReadingViews() {
```
The tests are complaining that this is not a function. This needs to be resolved. After that we can see if any are still failing.

huss reviewed (changes requested):
Thanks to @vyvyvyThao for another contribution. I've made some comments to address. Full testing is awaiting those changes.

Commit:
Update views refresh function and readings unit test cases

Commit:
update reading unit tests and view code

vyvyvyThao committed:
Add missing DELTA declaration

huss commented on src/server/services/refreshAllReadingViews.js:
```diff
@@ -6,15 +6,31 @@
 
 const { log } = require('../log');
 
```
Please remove blank line.

huss commented on src/server/services/refreshAllReadingViews.js:
```diff
@@ -6,15 +6,31 @@
+const { getConnection } = require('../db');
+const Reading = require('../models/Reading');
 
 
```
Please remove second blank line.

huss commented on src/server/services/refreshAllReadingViews.js:
```diff
@@ -6,15 +6,31 @@
+const Reading = require('../models/Reading');
 
 
+/*
```
Maybe make this /** since it is a function description. Also, remove blank line between the comment and the function declaration.

huss commented on src/server/services/refreshAllReadingViews.js:
```diff
@@ -6,15 +6,31 @@
+	const conn = getConnection();
+
+	// Refresh hourly readings view
 
```
OED prefers not to have a blank link after a comment so please remove here and below.

huss commented on src/server/services/refreshAllReadingViews.js:
```diff
@@ -6,15 +6,31 @@
+	await Reading.refreshDailyReadings(conn);
+	log.info('Daily View Refreshed');
 }
 
```
Please remove second blank line.

huss commented on src/server/services/refreshHourlyReadingViews.js:
```diff
@@ -9,6 +9,10 @@ const { log } = require('../log');
 const { getConnection } = require('../db');
 const Reading = require('../models/Reading');
 
+/** 
```
A couple of thoughts:

- I think the message would be better if it said why and what to use instead. Maybe: @deprecated: OED only supports refreshing all views so please use refreshAllReadingViews.
- Also, use @ see to link to other file.
- The comment should end with */ not **/.
- Please remove blank line between comment and function declaration.

huss commented on src/server/services/refreshReadingViews.js:
```diff
@@ -16,6 +16,11 @@ const Reading = require('../models/Reading');
  * Refreshes daily view.
  */
+
+/** 
```
Please see comments in refreshHourlyReadingView and apply similarly here.

huss commented on src/server/services/refreshAllReadingViews.js:
```diff
@@ -6,15 +6,31 @@
-	await Promise.all([refreshDailyReadingViews(), refreshHourlyReadingViews()]);
-	log.info('...Views Refreshed!');
+	log.info('Refreshing Materialized Hourly Reading Views');
+	await Reading.refreshHourlyReadings(conn);
```
I think it might make sense to create a new function in Reading that does all the refreshing and has this logic. This puts the logic in the model. It will also make it easier to change the refresh in the model for any future changes.

huss commented on src/server/test/db/unitReadingsTests.js:
```diff
@@ -79,7 +84,9 @@ mocha.describe('Line & bar Readings', () => {
 				new Reading(meter.id, 100, halfHourBefore, halfHourAfter)
 			], conn);
 
+			await Reading.refreshHourlyReadings(conn);
```
Per another comment, use the new Reading function to refresh all readings rather than doing each here. The code should be checked for any other places this should be done.

huss replied:
I now realize you don't need to do groups. However, per another comment, this will likely be a single function to do both. It is okay to do both if that is easier.

huss commented on src/server/test/db/unitReadingsTests.js:
```diff
@@ -15,6 +15,8 @@ const Conversion = require('../../models/Conversion');
 const { insertStandardUnits, insertStandardConversions } = require('../../util/insertData');
 const { insertSpecialUnits, insertSpecialConversions } = require('../../data/automatedTestingData');
 const { redoCik } = require('../../services/graph/redoCik');
+// Readings should be accurate to many decimal places, but allow some wiggle room for database and javascript conversions
```
Please put a blank line before the comment to separate from the imports.

huss commented on src/server/test/db/unitReadingsTests.js:
```diff
@@ -453,6 +462,7 @@ mocha.describe('Line & bar Readings', () => {
 			graphicUnitId = (await Unit.getByName('MJ', conn)).id;
 		});
 
+
```
Please remove second blank line added.

huss commented on src/server/test/db/unitReadingsTests.js:
```diff
@@ -562,12 +576,14 @@ mocha.describe('Line & bar Readings', () => {
+
+			for (let i = 0; i < readingsForMeterComparable.length; i++) {
+				expect(readingsForMeterComparable[i].reading).to.be.closeTo((100 + (i * 100)) * conversionSlope, DELTA);
+				expect(readingsForMeterComparable[i].start_timestamp).to.be.closeTo(start_timestamps[i], DELTA);
```
I'm unclear on what a delta on a timestamp does. Is it really needed?

huss replied:
I changed to be equal and it worked fine.

huss reviewed (changes requested):
Thanks to @vyvyvyThao for the updated code. I have made some comments. I resolved the previous comments that were done but there are a few that did not seemed to be addressed. Please let me know if you have any questions/thoughts/

vyvyvyThao committed:
group 3d function updates

vyvyvyThao committed:
Merge branch 'OpenEnergyDashboard:development' into views

vyvyvyThao committed:
Optimize group views

vyvyvyThao committed:
Minor db updates

vyvyvyThao committed:
Create refresh groups deep meters view npm command

vyvyvyThao committed:
Update groups api and edit group modal component

vyvyvyThao committed:
Automate refreshing group views after group edits

vyvyvyThao committed:
Merge branch 'OpenEnergyDashboard:development' into views

vyvyvyThao committed:
Update group compare

vyvyvyThao committed:
Merge branch 'views' of github.com:vyvyvyThao/OED into views

vyvyvyThao committed:
Test cases fix

vyvyvyThao committed:
Finallize pull request

vyvyvyThao committed:
Merge branch 'OpenEnergyDashboard:development' into views

vyvyvyThao committed:
Add updated views and functions for migration

vyvyvyThao committed:
Migration updates

huss committed:
Merge branch 'development' into pr/vyvyvyThao/1480

huss commented:
I merged development into timeVary which created a merge conflict. I pushed a commit to resolve that.

huss commented:
I changed the target branch to development to try to see if CodeQL still had an issue with the groups route given the new check. However, I could not figure out how to get them to run so I changed the target back to timeVary. I think OED needs to wait until timeVary is merged into development (in the future) to see if it is still an issue unless someone else has ideas.

vyvyvyThao committed:
Modify a global const

huss commented on src/server/services/refreshGroupsDeepMetersView.js:
```diff
@@ -0,0 +1,25 @@
+ */
+
+const { log } = require('../log');
+
```
Please remove blank line.

huss commented on src/server/models/Reading.js:
```diff
@@ -69,6 +69,22 @@ class Reading {
+		// TODO This will be removed once we completely transition to the unit version.
+		return conn.none('REFRESH MATERIALIZED VIEW hourly_readings_unit');
+	}
+
```
When I searched the code, I only found a few uses where it does daily without hourly. (See another comment about it that should change.) Given the new code setup, should the model have a function to do both in the correct order to be sure it is always done correctly. If so, maybe these can go away.

huss commented on src/server/models/Group.js:
```diff
@@ -225,6 +225,16 @@ class Group {
 
+	/**
+	 * Refreshes the groups deep meters view.
+	 * Should be called whenever a group's child meters are edited.
```
I think this should be "group's child meters or groups".

huss commented on src/server/routes/groups.js:
```diff
@@ -13,6 +13,9 @@ const { adminAuthMiddleware, optionalAuthMiddleware } = require('./authenticator
 const Point = require('../models/Point');
 const { failure, success } = require('./response');
+const { refreshGroupsDeepMetersView } = require('../services/refreshGroupsDeepMetersView');
+const { property } = require('lodash');
```
This import does not seem to be used.

huss commented on src/server/routes/groups.js:
```diff
@@ -297,6 +300,7 @@ router.put('/edit', adminAuthMiddleware('edit groups'), async (req, res) => {
 	const validGroup = {
 		type: 'object',
 		maxProperties: 10,
+		
```
Please remove new blank line.

huss commented on src/server/routes/groups.js:
```diff
@@ -350,49 +358,50 @@ router.put('/edit', adminAuthMiddleware('edit groups'), async (req, res) => {
+
 			res.sendStatus(200);
 		} catch (err) {
+			console.log("200", err);
```
remove console.log

huss commented on src/server/routes/groups.js:
```diff
@@ -403,6 +412,11 @@ router.put('/edit', adminAuthMiddleware('edit groups'), async (req, res) => {
 });
 
+router.post('/refresh', adminAuthMiddleware('refresh group views'), async (req, res) => {
+	await refreshGroupsDeepMetersView();
```
Though OED may not do this consistently (yet), it is probably good practice to use a try/catch where the catch returns the correct error code.

huss commented on src/server/test/db/unitReadingsTests.js:
```diff
@@ -449,10 +459,15 @@ mocha.describe('Line & bar Readings', () => {
+
 			// Make the graphic unit be MegaJoules.
 			graphicUnitId = (await Unit.getByName('MJ', conn)).id;
+
```
Can you please remove this blank line and the second one in two lines.

huss commented on src/server/test/db/unitReadingsTests.js:
```diff
@@ -586,33 +613,47 @@ mocha.describe('Line & bar Readings', () => {
-			]);
+
+			expect(readingsForMeterComparable[0].reading).to.be.closeTo(300 * conversionSlope, DELTA);
+			expect(readingsForMeterComparable[0].start_timestamp).to.be.closeTo(timestamp1.valueOf(), DELTA);
```
See comment above.

huss replied:
fixed per comment above

huss commented on src/server/test/db/unitReadingsTests.js:
```diff
@@ -15,6 +15,9 @@ const Conversion = require('../../models/Conversion');
 const { redoCik } = require('../../services/graph/redoCik');
+const { refreshGroupsDeepMetersView } = require('../../services/refreshGroupsDeepMetersView');
+// Readings should be accurate to many decimal places, but allow some wiggle room for database and javascript conversions
+const DELTA = 0.0000001;
```
src/server/util/readingsUtils.js already had DELTA defined so that should be used.

huss commented on src/server/test/web/groups.js:
```diff
@@ -191,9 +193,12 @@ mocha.describe('groups API', () => {
 						// login
 						res = await chai.request(app).post('/api/login')
 							.send({ username: unauthorizedUser.username, password: unauthorizedUser.password });
+						// console.log("Res:", res);
```
Please remove this and the next two console.log.

huss commented on src/server/test/web/readingsBarMeterQuantity.js:
```diff
@@ -40,6 +40,7 @@ mocha.describe('readings API', () => {
                             barWidthDays: 1,
                             graphicUnitId: unitId
                         });
+
```
Could you please remove this blank line since only change and no other test in file does that.

huss commented on src/server/util/globalConst.js:
```diff
@@ -0,0 +1,13 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+const MIN_ITEMS = 0;
```
I think a comment on what these are for would be good.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -350,6 +245,98 @@ CREATE EXTENSION IF NOT EXISTS btree_gist;
+	to all child meters in that group.
+*/
+CREATE OR REPLACE FUNCTION get_graphic_unit (
+	meters_group_id INTEGER
```
Should this be named group_id?

vyvyvyThao replied:
I think the different naming makes it not ambiguous when comparing it to gdm.group_id

huss replied:
I renamed it with the prefix requested_ similarly to other places to avoid name ambiguity. I had to drop the function with cascade due to the name change - only applies if had the old version.

huss commented on src/client/app/components/groups/EditGroupModalComponent.tsx:
```diff
@@ -336,9 +336,20 @@ export default function EditGroupModalComponent(props: EditGroupModalComponentPr
 					// This saves group to the DB and then refreshes the window if the last group being updated and
 					// changes were made to the children. This avoid a reload on name change, etc.
-					submitGroupEdits(submitState);
+					submitGroupEdits({ editedGroup: submitState, shouldRefreshGroupsDeepMetersView: shouldRefreshGroupsDeepMeters })
```
When I save a group with child meter changes (may happen at other times), there is an exception (can see in web browser console or via debugging). I'm also seeing that the refresh of the deep group view you added is not happening. I'm not sure what is up as I stopped debugging at that point. Does it work for you?

huss replied:
I cannot reproduce this since the last changes. I'm going to leave it unresolved for now but it may be fixed.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -350,6 +245,98 @@ CREATE EXTENSION IF NOT EXISTS btree_gist;
+group_daily_readings_unit
+	AS SELECT
+		gdm.group_id,
+		sum(dr.reading_rate  * c.slope + c.intercept) AS reading_rate,
```
Extra space before *.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -350,6 +245,98 @@ CREATE EXTENSION IF NOT EXISTS btree_gist;
+	GROUP BY gdm.group_id, gu.graphic_unit_id, hr.time_interval
+	ORDER BY gdm.group_id;
+
+CREATE INDEX if not exists idx_group_hourly_readings_unit ON group_hourly_readings_unit USING GIST(time_interval, group_id, graphic_unit_id);
```
Is there a reason this index has the last two items reversed from the daily index?

vyvyvyThao replied:
I think I reversed them for testing to see if it improved the runtime. I changed it back.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -679,12 +683,15 @@ BEGIN
 	-- end timestamp by that amount so it stops at the desired end timestamp.
 	real_end_stamp := real_end_stamp - bar_width;
 
+	RAISE NOTICE 'real_start_stamp: %, real_end_stamp: %, num_bars: %',
```
This should go.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -729,22 +736,51 @@ DECLARE
-	SELECT array_agg(DISTINCT gdm.meter_id) INTO meter_ids
-	FROM groups_deep_meters gdm
-	INNER JOIN unnest(group_ids) gids(id) ON gdm.group_id = gids.id;
+	bar_width := INTERVAL '1 day' * bar_width_days;
```
I have not thought it through but since this code is in both meter & group for bar could it be a function that is reused?

huss replied:
I checked this more carefully. The meter version sets real_tsrange in a different way. There are also two variables at the end that are needed elsewhere so the function needs to return two values. Could do with OUT parameters but given all of this decided not to do it.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -350,6 +245,98 @@ CREATE EXTENSION IF NOT EXISTS btree_gist;
+	child_meters_unit_ids INTEGER[];
+	unit_ids INTEGER[] := '{}';
+	unit_id INTEGER;
+	curr_src_id INTEGER;
```
I don't see this used.

huss commented on src/server/sql/reading/create_reading_views.sql:
```diff
@@ -350,6 +245,98 @@ CREATE EXTENSION IF NOT EXISTS btree_gist;
+	src_ids INTEGER[];
+	dest_ids INTEGER[];
+	child_meters_unit_ids INTEGER[];
+	unit_ids INTEGER[] := '{}';
```
I dislike variable names that only vary by small amount, esp. if only one character at the end. I think it is prone to mistakes. Could you please change this.

huss replied:
I renamed to unit_ids_compatible.

huss commented on src/server/migrations/1.0.0-2.0.0/sql/readings/create_reading_views.sql:
```diff
@@ -2,138 +2,97 @@
+/*
+Rounds a timestamp up to the next interval
+ */
+CREATE OR REPLACE FUNCTION date_trunc_up(interval_precision TEXT, ts TIMESTAMP)
```
This is almost a copy of the sql/ one. This should only have functions/views that contain a change that impacts how it works (for example, not doing SUM -> sum unless there are other changes that matter).

huss commented on src/server/migrations/1.0.0-2.0.0/sql/readings/create_reading_views.sql:
```diff
@@ -2,138 +2,97 @@
+so the line reading functions can use them both in the same way.
+ */
 
 CREATE MATERIALIZED VIEW IF NOT EXISTS
```
For whatever views are left, it should not have ``IF NOT EXISTS`` because then it won't update those views on migration. The new code in sql/ would update the next time there is a refresh. I realize I probably did this one so please change all that are needed.

huss commented on src/server/sql/reading/create_function_get_compare_readings.sql:
```diff
@@ -119,29 +119,68 @@ shift: How far back in time to shift the curr_start and curr_end date/time to ge
 )
 	RETURNS TABLE(group_id INTEGER, curr_use FLOAT, prev_use FLOAT)
 AS $$
+-- DECLARE
```
Can this go at this point?

huss commented on src/server/sql/reading/create_function_get_compare_readings.sql:
```diff
@@ -119,29 +119,68 @@ shift: How far back in time to shift the curr_start and curr_end date/time to ge
 END;
 $$ LANGUAGE 'plpgsql';
+
+
```
Can you please remove the extra two blank lines.

huss commented on src/server/sql/reading/create_function_get_3d_readings.sql:
```diff
@@ -158,9 +158,6 @@ BEGIN
-/* Gets group meters graphing data for 3D graphic by returning points that span the requested
-  length of time over the days requested. 
-*/
 CREATE OR REPLACE FUNCTION group_3d_readings_unit (
```
Could you help me understand why this comment was removed:
```
/* Gets group meters graphing data for 3D graphic by returning points that span the requested
  length of time over the days requested. 
*/
```

huss commented on src/server/migrations/1.0.0-2.0.0/sql/group/create_groups_tables.sql:
```diff
@@ -0,0 +1,47 @@
+  TODO: Deal with parent meters that are installed after their children. They only shadow them from a start-date onwards.
+  The above to-do is probably going to require a significant reworking of some stuff.
+ */
+CREATE MATERIALIZED VIEW IF NOT EXISTS groups_deep_meters AS
```
I'm curious what happens to the previous view that existed with the same name.  I think it may need to be dropped to do this but have not tested.

vyvyvyThao replied:
I manually dropped it on cascade and so everything is created again when running OED

huss replied:
I added the drop of the table.

huss commented on src/client/app/redux/api/groupsApi.ts:
```diff
@@ -9,13 +9,16 @@ import { GroupChildren, GroupData } from '../../types/redux/groups';
 import { selectIsAdmin } from '../slices/currentUserSlice';
 import { baseApi } from './baseApi';
+import { setRefreshingReadings } from '../../redux/slices/appStateSlice';
+
```
remove extra blank line.

huss commented on src/client/app/redux/api/groupsApi.ts:
```diff
@@ -9,13 +9,16 @@ import { GroupChildren, GroupData } from '../../types/redux/groups';
 });
 export const groupsInitialState = groupsAdapter.getInitialState();
 export type GroupDataState = EntityState<GroupData, number>;
 
```
remove extra blank line.

huss commented on src/client/app/redux/api/groupsApi.ts:
```diff
@@ -104,3 +132,5 @@ export const selectGroupNameWithID = (state: RootState, groupId: number) => {
 	return groupInfo ? groupInfo.name : '';
 };
+
+export const stableEmptyGroups: GroupData[] = [];
```
I don't see this used anywhere. What is it for?

huss commented on src/client/app/components/groups/GroupsDetailComponent.tsx:
```diff
@@ -21,45 +25,57 @@ export default function GroupsDetailComponent() {
 	// We only want displayable groups if non-admins because they still have non-displayable in state.
 	const { visibleGroups } = useAppSelector(state => selectVisibleMeterAndGroupData(state));
-
+	selectGroupDataResult;
```
I'm unclear on what this is for as it isn't in an assignment, etc. I think I'm missing something.

huss commented on src/client/app/redux/api/groupsApi.ts:
```diff
@@ -67,14 +70,39 @@ export const groupsApi = baseApi.injectEndpoints({
+			},
 			invalidatesTags: ['GroupData', 'GroupChildrenData']
 		}),
+		refreshGroups: builder.mutation<void, unknown>({
```
There seems to be other places where the second item in void instead of unknown. Could that be used and also avoid sending null on the call?

huss commented on src/client/app/redux/api/groupsApi.ts:
```diff
@@ -67,14 +70,39 @@ export const groupsApi = baseApi.injectEndpoints({
+				body: unknown
+			}),
+
+			// Only the group readings really need invalidation
```
Is this correct as the code invalidates other tags?

huss commented on src/client/app/redux/api/groupsApi.ts:
```diff
@@ -67,14 +70,39 @@ export const groupsApi = baseApi.injectEndpoints({
+					dispatch(groupsApi.endpoints.refreshGroups.initiate(null));
+				}
+			},
 			invalidatesTags: ['GroupData', 'GroupChildrenData']
```
In conversionApi.ts it does
```
} else {
	dispatch(conversionsApi.util.invalidateTags(['ConversionDetails']));
}
```
Is it different here for a reason? Is one preferred?

huss reviewed (changes requested):
Thanks to @vyvyvyThao for this substantial update to the PR based on OED's needs. Overall, it is nice. I have made a number of comments to address. Also, some previous comments are still unresolved. I did try to resolve any that were complete. I ask that you not resolve any comments so I can track which I have not rechecked; you are welcome to add a reply. Here are three other comments that I could not add where I wanted them:

- ~In src/server/services/eGauge/updateEgaugeMeters.js it does daily and then hourly which seems backward to me. This will possibly be replaced with a single call.~
- ~In src/server/test/db/unitReadingsTests.js it has a few places that don't do both refresh hour & daily. I can see it might save a little time if you know are only looking at one view in the test but I think it would be fine to do both if that allows for a single function in the model (see other comment).~
- ~It might be good to update the PR description for the new group page changes and that the limitation does not still seem to apply.~

I did modest testing to see if the code seems to work. It is also passing all CI tests that do not automatically run on this branch. I'll do more when this is close to complete. Please let me know if anything is not clear or if you have any thoughts.

dnna2610 committed:
Fix pull request

huss commented:
I'm noting that this work has been merged and used in the upcoming time-varying conversion work by @tdthompson650. That is now passing almost all the CI tests so this work should also be okay. I'm noting that any failed CI tests are not being addressed in this PR but other ones.

huss commented on src/server/util/readingsUtils.js:
```diff
@@ -40,7 +42,16 @@ async function prepareTest(unitData, conversionData, meterData, groupData = [])
+
+    // Only refresh meter views if there is no group changes.
+    if (groupData.length == 0) {
+        await Reading.refreshHourlyReadings(conn);
```
Should this use the new refreshMeterReadingsViews?

huss commented on src/server/test/db/unitReadingsTests.js:
```diff
@@ -494,6 +502,7 @@ mocha.describe('Line & bar Readings', () => {
 			// Associate both meters with a single group
```
Above this line and in one other place in the file seem to be he only places in testing that still use refreshHourlyReadings. Should it be changed to the new refreshMeterReadingsViews?

huss reviewed (changes requested):
Thanks to @vyvyvyThao for the update. I made a couple of new comments to address. Also, I resolved all the comments I thought were done but I still see a number of unresolved ones. Can you please address these and/or put in a comment about what you think. I will resolve them as needed.

vyvyvyThao committed:
Fix PR

vyvyvyThao committed:
Fix PR

huss committed:
fixes/changes

- fix function name missing s
- fix leftover line from removing raise notice
- fix import so js
- Wording on log msgs
- minor edits

huss committed:
resolve most remaining PR comments

- updated comment
- drop table to created in migration
- changed some variable names
- used equal for date testing

huss committed:
make migrations match new SQL

huss committed:
fuse group reading views update into single function

huss committed:
update state on group create

This fixes an issue where the groups were not refreshed so the deep meters
we not set on the server when a group was created. Create now mirrors
edit for groups.

huss committed:
modify group 3D to use group hourly view

This changed group 3D that used to used to sum the meter 3D results to directly
get the values from the new group hourly view. The idea is this is faster since it does
not sum over the meters.

As part of doing this, a function was created to determine the time for each point
and used in both the meter and group 3D case.

An index was added to mirror the one for hourly but added the new graphic unit.
A TODO was left since there are now two indices on the group hourly view and
it is unclear which are best/needed but adding this done substantially sped up the
3D data acquisition.

This changes the returned value when the request is inconsistent with 3D. For example,
the underlying meter frequency is too long. Previously it returned  -999 * # meters in
group but now it returns -999. This does not impact the code as it looks
for a negative value in this case. One test case was updated to reflect this.

huss commented on src/server/routes/groups.js:
```diff
@@ -349,47 +350,51 @@
 				await newGroup.update(t);
 
-				const adoptedGroups = difference(req.body.childGroups, currentChildGroups);
+				const adoptedGroups = difference(editedGroup.childGroups, currentChildGroups);
```
This is going to be ignored for the following reasons:

- The route is restricted to admins and OED generally trusts them. An admin would have to directly route to the OED server with bad values to cause any issue.
- The param checking restricts the size of the item so it cannot cause a major issue.

huss commented on src/server/routes/groups.js:
```diff
@@ -349,47 +350,51 @@
 
 				// Compute meters differences and adopt/disown to make changes
-				const adoptedMeters = difference(req.body.childMeters, currentChildMeters);
+				const adoptedMeters = difference(editedGroup.childMeters, currentChildMeters);
```
This is going to be ignored for the following reasons:

- The route is restricted to admins and OED generally trusts them. An admin would have to directly route to the OED server with bad values to cause any issue.
- The param checking restricts the size of the item so it cannot cause a major issue.

huss commented:
All comments have been addressed. The code is passing all checks/tests (done on local machine since not automatically done on this branch). Modest testing indicates it is working properly. This is acceptable since it is not going into the development branch at this time. As previously noted, this is used in another PR that needs this code so merging is desirable. Further testing will be done as part of that PR and in the future. Thank you to @vyvyvyThao for doing this work.
